hw/ip/otbn/util/shared/constants.py

```diff
@@ -46,8 +47,15 @@ def get(self, gpr: str) -> Optional[int]:
     def __contains__(self, gpr: str) -> bool:
         return gpr in self.values
 
-    def copy(self) -> 'ConstantContext':
-        return ConstantContext(self.values)
+    def __deepcopy__(self, memo) -> 'ConstantContext':
+        return ConstantContext(deepcopy(self.values, memo))
```

Comment:
Can you justify this advice? This seems like a completely natural approach to me. In general, advice about code structure that isn't obvious should be acompanied by justification.

Reply:
Well, taking copies like this usually shows that you're mutating state by accident. Which isn't great! I guess you *can* take copies of everything everywhere and pretend you're using a very slow Haskell implementation, but it's not really how Python programs normally expect to work.

My opinion is that it's better to be explicit about what is being mutated and what isn't and do updates in methods (to avoid the code becoming spaghetti).

Reply:
Actually, I didn't just implement the deepcopy because some state was unexpectedly mutated. There were a few specific places where I knew I needed to fully copy certain state objects (the constants or the information-flow graphs). And when I needed to do that, I chose a `deepcopy` method simply because it has existing recognition and conveyed the exact thing I wanted to do, so I thought it would be more understandable than creating a method called something else.

The reasons I needed the deepcopies were, for example: I copy the constant context that's passed into `get_iflow` because I want a version of that which I can update as I traverse the control-flow graph, but don't want to modify the context that the caller passed me. So I need to duplicate it. I could have just called it `copy` (since there's only one level here anyway, really) but I wanted to be clear that it was safe to assume no shared pointers.